plugin/pkg/admission/imagepolicy/admission.go

```diff
@@ -0,0 +1,255 @@
+}
+
+// AdmissionConfig holds config data for admission controllers
+type AdmissionConfig struct {
+	Webhook imagePolicyWebhookConfig `json:"webhook"`
+}
+
+func (a *imagePolicyWebhook) statusTTL(status v1beta1.ImageReviewStatus) time.Duration {
+	if status.Allowed {
+		return a.allowTTL
+	}
+	return a.denyTTL
+}
+
+// Filter out annotations that don't match *.image-policy.k8s.io/*
+func (a *imagePolicyWebhook) filterAnnotations(allAnnotations map[string]string) map[string]string {
+	annotations := make(map[string]string)
+	for k, v := range allAnnotations {
+		if strings.Contains(k, ".image-policy.k8s.io/") {
+			annotations[k] = v
+		}
+	}
+	return annotations
+}
+
+// Function to call on webhook failure; behavior determined by defaultAllow flag
+func (a *imagePolicyWebhook) webhookError(err error) error {
+	if err != nil {
+		glog.V(4).Infof("error contacting webhook backend: %s")
+		if a.defaultAllow {
+			glog.V(4).Infof("pod allowed in spite of webhook backend failure")
+			return nil
+		}
+		glog.V(4).Infof("pod not allowed due to webhook backend failure ")
+		return err
+	}
+	return nil
+}
+
+func (a *imagePolicyWebhook) Admit(attributes admission.Attributes) (err error) {
+	// Ignore all calls to subresources or resources other than pods.
+	if len(attributes.GetSubresource()) != 0 || attributes.GetResource().GroupResource() != api.Resource("pods") {
+		return nil
+	}
+	pod, ok := attributes.GetObject().(*api.Pod)
+	if !ok {
+		return apierrors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
+	}
+
+	annotations := a.filterAnnotations(pod.Annotations)
+	namespace := attributes.GetNamespace()
+
+	// Build list of ImageReviews to send to backend, one per image
+	var imageReviews []v1beta1.ImageReview
+	for _, c := range pod.Spec.Containers {
+		imageReviews = append(imageReviews, v1beta1.ImageReview{
+			Spec: v1beta1.ImageReviewSpec{
+				Container: v1beta1.ImageReviewContainerSpec{
+					Image: c.Image,
+				},
+				Annotations: annotations,
+				Namespace:   namespace,
+			},
+		})
+	}
+
+	for _, r := range imageReviews {
+		if err := a.admitContainer(&r); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (a *imagePolicyWebhook) admitContainer(review *v1beta1.ImageReview) error {
+	cacheKey, err := json.Marshal(review.Spec)
+	if err != nil {
+		return err
+	}
+	if entry, ok := a.responseCache.Get(string(cacheKey)); ok {
+		review.Status = entry.(v1beta1.ImageReviewStatus)
+	} else {
+		result := a.WithExponentialBackoff(func() restclient.Result {
+			return a.RestClient.Post().Body(review).Do()
+		})
+
+		if err := result.Error(); err != nil {
+			return a.webhookError(err)
+		}
+		var statusCode int
+		if result.StatusCode(&statusCode); statusCode < 200 || statusCode >= 300 {
+			return a.webhookError(fmt.Errorf("Error contacting webhook: %d", statusCode))
+		}
+
+		raw, err := result.Raw()
+		if err != nil {
+			return a.webhookError(err)
+		}
+		json.Unmarshal(raw, review)
```

Comment:
This only exists because the registration for the current API group seems to be broken. We're still trying to hunt down why it doesn't recognize the `ImageReview` type.